core/src/main/java/com/linecorp/armeria/common/stream/FlatMapStreamMessage.java

```diff
@@ -0,0 +1,378 @@
+                                     SubscriptionOption... options) {
+            requireNonNull(downstream, "downstream");
+            requireNonNull(function, "function");
+            requireNonNull(executor, "executor");
+            requireNonNull(completionFuture, "completionFuture");
+
+            this.downstream = downstream;
+            this.function = function;
+            this.executor = executor;
+            this.maxConcurrency = maxConcurrency;
+            this.completionFuture = completionFuture;
+            this.options = options;
+
+            childSubscribers = new HashSet<>();
+            buffer = new ArrayDeque<>();
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            requireNonNull(subscription, "subscription");
+
+            upstream = subscription;
+            downstream.onSubscribe(this);
+        }
+
+        @Override
+        public void onNext(T item) {
+            requireNonNull(item, "item");
+
+            if (canceled) {
+                StreamMessageUtil.closeOrAbort(item);
+                return;
+            }
+
+            pendingSubscriptions++;
+            final StreamMessage<U> newStreamMessage = function.apply(item);
+            newStreamMessage.subscribe(new FlatMapSubscriber<>(this), executor, options);
+        }
+
+        @Override
+        public void onError(Throwable cause) {
+            requireNonNull(cause, "cause");
+
+            if (canceled) {
+                return;
+            }
+            canceled = true;
+
+            cancelChildSubscribers();
+            downstream.onError(cause);
+        }
+
+        @Override
+        public void onComplete() {
+            if (canceled) {
+                return;
+            }
+
+            if (childSubscribers.isEmpty() && pendingSubscriptions == 0) {
+                downstream.onComplete();
+                completionFuture.complete(null);
+            } else {
+                completing = true;
+            }
+        }
+
+        @Override
+        public void request(long n) {
+            if (n <= 0) {
+                onError(new IllegalArgumentException(
+                        "n: " + n + " (expected: > 0, see Reactive Streams specification rule 3.9)"));
+                cancel();
+                return;
+            }
+
+            if (canceled) {
+                return;
+            }
+
+            if (executor.inEventLoop()) {
+                handleRequest(n);
+            } else {
+                executor.execute(() -> handleRequest(n));
+            }
+        }
+
+        private void handleRequest(long n) {
+            requestedByDownstream = LongMath.saturatedAdd(requestedByDownstream, n);
+            flush();
+
+            final long toRequest = maxConcurrency - childSubscribers.size();
+            if (toRequest > 0) {
+                upstream.request(toRequest);
+            }
```

Comment:
It seems like we should also request upstream once a subscriber completes and is removed from `childSubscribers`.
e.g. 3 child subscribers with 2 maxConcurrency may result in the third child not processing.

```
final StreamMessage<Integer> streamMessage = StreamMessage.of(10, 20, 30);
final Function<Integer, StreamMessage<Integer>> function =
        num -> StreamMessage.of(Flux.range(num, 3).delayElements(Duration.ofSeconds(1)));
final StreamMessage<Integer> mappedStream = streamMessage.flatMap(function, 2);

StepVerifier.create(mappedStream)
            .thenRequest(100L)
            .thenConsumeWhile(num -> {
                System.out.println("num: " + num);
                return true;
            })
            .verifyComplete();
```

Reply:
To give a more simple example of the deadlock example I mentioned earlier:
A child stream may not emit any elements and complete the its stream.
In this case, although there are outstanding requests it may be possible that the stream doesn't emit any more elements and a deadlock occurs.

Related to my comment above, I wonder if we want to just request upstream once and be over with it.

```suggestion
            if (!initialized) {
                initialized = true;
                upstream.request(maxConcurrency);
            }
```

and request upstream each time a child completes to maintain `maxConcurrency` count of child streams.


Also as it stands, since child streams are added when `onSubscribe` is invoked it is possible that >3 child streams are subscribed to although `maxConcurrency == 3`

Reply:
Good point, I think that would simplify this part a lot.

>Also as it stands, since child streams are added when onSubscribe is invoked it is possible that >3 child streams are subscribed to although maxConcurrency == 3

I don't understand this part. If you mean `FlatMapSubscriber.onSubscribe`, it does add more child subscribers but the number of `FlatMapSubscriber.onSubscribe` calls is limited by the number of elements requested from the upstream, which is in turn limited by `maxConcurrency`.

Reply:
I thought the following scenario was possible (let me know if I'm misunderstanding)

Assuming a `maxConcurrency == 3`
1. FlatMapAggregatingSubscriber.request(3)
2. FlatMapAggregatingSubscriber.request(3) 
4. FlatMapAggregatingSubscriber.onSubscribe() is invoked for 6 elements
5. childSubscribers.size() == 6

Reply:
Ah, you're right 😅 If multiple requests are made before any of the inner subscriptions are created it will make multiple requests to upstream. This is neatly fixed by the suggestion you made with an initialization flag 👍